### ISSUE TITLE
skip: true and skipToken behave differently with useSuspenseQuery

### ISSUE THREAD

**Issue by epmatsw** [apollographql/apollo-client#12059]
### Issue Description

I'm pretty sure this is a bug with `skip: true`, or at least `skipToken` behaves like I'd expect.

If you have a `useSuspenseQuery` that uses `skip: boolean` and toggles between true and false for that, it *seems* like sometimes the switch from true to false causes the query to skip the cache and suspend.

In the demo, you can quickly hit the button to toggle skipping on or off. Every once in a while, you can see the top one (skip: boolean) pop a suspend for a few frames, while the bottom one (skipToken) correctly re-reads from the cache.

### Link to Reproduction

https://codesandbox.io/p/sandbox/lz67k5?file=%2Fsrc%2FApp.tsx%3A34%2C67

### `@apollo/client` version

3.11.8

**Comment by jerelmiller** [comment]
This is such an interesting one. Both `skip: true` and `skipToken` have almost the same code paths. It looks like the only difference is that we set some extra options for `skip: true` (not sure why to be honest). See how `skipToken` returns early while `skip: true` adds the extra options: https://github.com/apollographql/apollo-client/blob/1f0333e1644cb95b6862441ff957e60ec5f7a59e/src/react/hooks/useSuspenseQuery.ts#L357-L382

Other than this, both of these code paths rely on checking `fetchPolicy: 'standby'` in the downstream code which should avoid suspending (the `__use` hook is the thing that causes it to suspend): https://github.com/apollographql/apollo-client/blob/1f0333e1644cb95b6862441ff957e60ec5f7a59e/src/react/hooks/useSuspenseQuery.ts#L253

I'll have to dig into this further to understand where the difference is, but unfortunately won't have time to do so for a couple days. Thanks for raising the issue!

**Comment by naman1608** [comment]
Not sure but I think the `options.skip` check should also be moved with the `skipToken` check. It would help with the following - 
1. Consistency between both `options.skip` and `skipToken`
2. There is no need to add `notifyOnNetworkStatusChange` and 'nextFetchPolicy` as the query isn't being executed in the standby mode, so those lines seem redundant in case of `options.skip` as `true` also.

@jerelmiller I can put a PR if it makes sense

**Comment by jerelmiller** [comment]
@naman1608 you're welcome to try a fix! I think moving the check for `options.skip` up to the same conditional as `options === skipToken` makes sense as a starting point. The one difference here is that `variables` is set in `options` with `options.skip`, so we might want to add a test case to make sure that changing variables while `skip` is `true`, then changing `skip` to `false` properly suspends and fetches with the right variables (does that make sense?).

We are preparing for our upcoming 3.12 release, so if you end up getting a fix in place, we'll look at adding that to a 3.12 patch release. No obligation though!

**Comment by naman1608** [comment]
Yes, that makes sense. It would cause problem when using a reactive value as skip. Like - 
```
const [skip, setSkip] = useState(true);

const { data } = useSuspenseQuery(query, {
  skip,
  variables: { id: '123' },
  fetchPolicy: 'network-only'
});

// Later...
setSkip(false); // Problem: We'd lose variables and other options if we moved options.skip check to the top
```
I think the main thing is how skipToken and options.skip are perceived. As in, skipToken seems to have a purpose of completely bypassing the query execution, when we don't have the variables yet.
Whereas with options.skip, it seems like it is used to temporarily disable the execution, with the variables may or may not be set.

**Comment by jerelmiller** [comment]
@naman1608 by the way, I don't want you to feel obligated to fixing this. By all means, if you want to contribute a fix, absolutely go for it! Otherwise we'll try and tackle this sometime after we release 3.12 🙂.

**Comment by jay05410** [comment]
Hi @jerelmiller, I see this issue is still open and I’ve been following the discussion. I’ve analyzed the code path in useSuspenseQuery.ts and noticed the inconsistency you mentioned between skipToken and options.skip.

I’m very interested in working on this to unify the logic and prevent unnecessary suspensions.
Please let me know if it's available for contribution. Thank you!

**Comment by jerelmiller** [comment]
@jay05410 go for it! To clarify, the goal isn't necessarily to unify the logic, but rather to figure out why the skip behavior is different between the two causing the component to suspend.

I'm realizing now the reason why we add extra options with `skip: true` is because we want to allow you to `refetch` when `skip` is `true` much like `useQuery` does: https://github.com/apollographql/apollo-client/blob/9cbe2c2dd2282ac861327d3c394578db7706df05/src/react/hooks/__tests__/useQuery.test.tsx#L8534-L8582

Apologies for adding confusion with my earlier statement!

We want `skip: true` to apply as many options as possible to the underlying `ObservableQuery` so that `refetch` uses the right options (such as `variables`). We also want to allow `refetch` with `skipToken`, but the difference obviously is that we don't have an options object provided to the hook. You'll see a similar test in `useQuery` for `refetch` with `skipToken`: https://github.com/apollographql/apollo-client/blob/9cbe2c2dd2282ac861327d3c394578db7706df05/src/react/hooks/__tests__/useQuery.test.tsx#L8584-L8629

Looking through the `useSuspenseQuery` test cases, it seems we don't have these tests, so perhaps you might be willing to add one as well if you tackle this?
